design/multiple-csi-volumesnapshotclass-support.md

```diff
@@ -0,0 +1,216 @@
+
+## Goals
+- Allow the user to specify the VolumeSnapshotClass to use for a particular driver and backup.
+
+## Non Goals
+-
+
+
+## Detailed Design
+
+### Plugin Inputs Contract Changes
+Approach is to introduce a new field `pluginInputs` in the velero CRs (Backup, Schedule, Restore). This field can be leveraged by all plugins for sending plugin specific settings rather than relying on annotations or global settings which hold across backups.
+
+```go
+type PluginInput struct {
+    Name string `json:"name"`
+    Properties map[string][string] `json:"properties"`
+}
+```
+
+### Using Plugin Inputs for CSI Plugin
+The user can specify the VolumeSnapshotClass to use for a particular driver and backup using the plugin inputs. The CSI plugin will use the VolumeSnapshotClass specified in the plugin inputs. If the VolumeSnapshotClass is not specified for a driver, the CSI plugin will use the default VolumeSnapshotClass for the driver fetched using labels through older route.
+
+Example: 
+```yaml
+apiVersion: velero.io/v1
+kind: Backup
+metadata:
+  name: backup-1
+spec:
+    pluginInputs:
+    - name: velero.io/csi
+    - properties:
+        - key: csi.cloud.disk.driver
+        - value: csi-diskdriver-snapclass
+        - key: csi.cloud.file.driver
+        - value: csi-filedriver-snapclass
+```
+
+CLI Example
+
+```bash
+velero backup create my-backup --plugin-inputs velero.io/csi:csi.cloud.disk.driver=csi-diskdriver-snapclass,csi.cloud.file.driver=csi-filedriver-snapclass
+```
+
+### Annotations overrides on PVC for CSI Plugin
+The user can annotate the PVCs with VolumeSnapshotClass name. This will override whatever the user has passed in pluginInputs for that driver.
+
+- If annotation is not present or VolumeSnapshotClass referred is not present in cluster OR if the specified VSC does not have the same CSI driver as the PVC
+    -  the CSI plugin will try to fallback to the pluginInputs value for that driver. If pluginInputs does not have the VSC for that driver, the CSI plugin will use the default VolumeSnapshotClass for the driver using the older label route.
+
+Example: 
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+name: pvc-1
+annotations:
+    velero.io/csi-volumesnapshot-class/disk.fg: csi-diskdriver-snapclass
+```
+
+### Using Plugin Inputs for Other Plugins
+As of today various plugins such as StorageClass mapping plugin, use a global configmap which is discovered through labels/ annotations which is not ideal customer experience and prohibhit usage of specific settings for each backup/ schedule setup.
+
+The pluginInputs field can be used to pass in the reference the configmap/ secret to use for a particular plugin. This will allow the user to use different configmaps/ secrets for different backups/ schedules.
```

Comment:
What if both the global cm as well as the pluginInputs cm exist ?

Reply:
Similar to the CSI driver case
where we use pluginInputs where available and fallback on global cm for other cases.
Guidance will be such that all plugins should try to honour first the params in pluginInput(be it explicit params such as VSC, or pointer to CM) if it is not able to get required info from pluginInput, fallback will be global CM

I will add this guidance.